platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/merkle/AbstractHashListener.java

```diff
@@ -97,78 +95,57 @@ protected AbstractHashListener(
-     */
-    @Override
-    public void onBatchStarted() {
-        batchLeaves.clear();
-        batchNodes.clear();
-    }
-
-    /**
-     * {@inheritDoc}
-     */
     @Override
-    public void onRankStarted() {
-        rankLeaves = new ArrayList<>(INITIAL_BATCH_ARRAY_SIZE);
-        rankNodes = new ArrayList<>(INITIAL_BATCH_ARRAY_SIZE);
+    public void onHashingStarted() {
+        nodes = new ArrayList<>();
+        leaves = new ArrayList<>();
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
     public void onNodeHashed(final long path, final Hash hash) {
-        rankNodes.add(new VirtualHashRecord(path, hash));
-    }
-
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public void onLeafHashed(final VirtualLeafRecord<K, V> leaf) {
-        rankLeaves.add(leaf);
+        final List<VirtualHashRecord> dirtyHashesToFlush;
+        final List<VirtualLeafRecord<K, V>> dirtyLeavesToFlush;
+        synchronized (this) {
+            nodes.add(new VirtualHashRecord(path, hash));
+            if (nodes.size() > 10000) { // TODO: make it configurable
+                dirtyHashesToFlush = nodes;
+                nodes = new ArrayList<>();
+                dirtyLeavesToFlush = leaves;
+                leaves = new ArrayList<>();
+            } else {
+                dirtyHashesToFlush = null;
+                dirtyLeavesToFlush = null;
+            }
+        }
+        if ((dirtyHashesToFlush != null) && (dirtyLeavesToFlush != null)) {
+            flush(dirtyHashesToFlush, dirtyLeavesToFlush);
+        }
     }
 
-    /**
-     * {@inheritDoc}
-     */
     @Override
-    public void onRankCompleted() {
-        batchLeaves.add(rankLeaves);
-        batchNodes.add(rankNodes);
+    public synchronized void onLeafHashed(final VirtualLeafRecord<K, V> leaf) {
+        leaves.add(leaf);
     }
 
-    /**
-     * {@inheritDoc}
-     */
     @Override
-    public void onBatchCompleted() {
-        long maxPath = -1;
-
-        Stream<VirtualHashRecord> sortedDirtyHashes = Stream.of();
-        for (int i = batchNodes.size() - 1; i >= 0; i--) {
-            final List<VirtualHashRecord> batch = batchNodes.get(i);
-            if (!batch.isEmpty()) {
-                sortedDirtyHashes = Stream.concat(sortedDirtyHashes, batch.stream());
-                maxPath = Math.max(maxPath, batch.get(batch.size() - 1).path());
-            }
-        }
-
-        Stream<VirtualLeafRecord<K, V>> sortedDirtyLeaves = Stream.of();
-        for (int i = batchLeaves.size() - 1; i >= 0; i--) {
-            final List<VirtualLeafRecord<K, V>> batch = batchLeaves.get(i);
-            if (!batch.isEmpty()) {
-                sortedDirtyLeaves = Stream.concat(sortedDirtyLeaves, batch.stream());
-                maxPath = Math.max(maxPath, batch.get(batch.size() - 1).getPath());
-            }
+    public synchronized void onHashingCompleted() {
+        if (!nodes.isEmpty() || !leaves.isEmpty()) {
+            flush(nodes, leaves);
```

Comment:
At line 124 you call flush() outside of the `synchronized (this)` block, while here you call it implicitly inside this block because the method is synchronized. I think it would be safer if the synchronization contract for this method is the same for every call. This would ensure that we take a lock if one is required, or otherwise, avoid locking if it's unnecessary for running this method. I see that the method performs some I/O, so I'd guess you may want to release the lock before calling it. If my guess is wrong though, then it would help to add a comment in code here explaining why the lock is needed here specifically.

Reply:
I see no requirements for flush() to be called inside a synchronized block or not, other than performance. In `onNodeHashed()` it is called without synchronization to release the monitor (shared across all hashing tasks, as all of them use the same hashing listener) as soon as possible. In `onHashingCompleted()` this is not an issue, since hashing is completed, and no hashing tasks are running. I can change it to be consistent with `onNodeHashed()`, though...

Reply:
Done